### DISCUSSION TITLE
ip range not wok in rules

### DISCUSSION THREAD

**Discussion by mokhtarabadi** [eycorsican/leaf#53] [Q&A] [answered]
@eycorsican, what is wrong with my JSON config?

```
{
  "log": {
    "level": "info"
  },
  "inbounds": [
    {
      "protocol": "socks",
      "address": "127.0.0.1",
      "port": 1080,
      "tag": "socks_in"
    }
  ],
  "outbounds": [
    {
      "protocol": "chain",
      "settings": {
        "actors": [
          "vmess_tls",
          "vmess_ws",
          "vmess"
        ]
      },
      "tag": "vmess_out"
    },
    {
      "protocol": "tls",
      "settings": {
        "serverName": "example.org",
        "alpn": [
          "http/1.1"
        ]
      },
      "tag": "vmess_tls"
    },
    {
      "protocol": "ws",
      "settings": {
        "path": "/ray"
      },
      "tag": "vmess_ws"
    },
    {
      "protocol": "vmess",
      "settings": {
        "address": "example.org",
        "port": 443,
        "uuid": "b831381d-6324-4d53-ad4f-8cda48b30811",
        "security": "chacha20-ietf-poly1305"
      },
      "tag": "vmess"
    },
    {
      "protocol": "direct",
      "tag": "direct_out"
    }
  ],
  "rules": [
    {
      "ip": [
        "0.0.0.0/8",
        "10.0.0.0/8",
        "100.64.0.0/10",
        "127.0.0.0/8",
        "169.254.0.0/16",
        "172.16.0.0/12",
        "192.0.0.0/24",
        "192.0.2.0/24",
        "192.168.0.0/16",
        "198.18.0.0/15",
        "198.51.100.0/24",
        "203.0.113.0/24",
        "::1/128",
        "fc00::/7",
        "fe80::/10"
      ],
      "target": "direct_out"
    }
  ]
}
```

this config must redirect local IP ranges to direct_out but all IPs goes through vmess_out

**Comment by mokhtarabadi** [comment, marked as answer]
it works
the problem is firefox when "Proxy DNS when using SOCKS v5" is enabled, firefox recognized 192.168.1.1 as domain, not IP address

**Comment by nahddw** [comment]
{
  "log": {
    "level": "info"
  },
  "dns": {
    "servers": [
      "1.1.1.1",
      "8.8.8.8"
    ]
  },
  "inbounds": [
    {
      "address": "127.0.0.1",
      "port": 1088,
      "protocol": "socks"
    }
  ],
  "outbounds": [
    {
      "protocol": "failover",
      "settings": {
        "actors": [
          "vmess_out"
        ]
      },
      "tag": "failover_out"
    },
    {
      "protocol": "chain",
      "settings": {
        "actors": [
          "vmess_tls",
          "vmess_ws",
          "vmess"
        ]
      },
      "tag": "vmess_out"
    },
    {
      "protocol": "tls",
      "settings": {
        "serverName": "xxxx",
        "alpn": ["http/1.1"]
      },
      "tag": "vmess_tls"
    },
    {
      "protocol": "ws",
      "settings": {
        "path": "/path",
        "headers": {
          "Host": "xx.xx.xx.x"
        }
      },
      "tag": "vmess_ws"
    },
    {
      "protocol": "vmess",
      "settings": {
        "address": "xx.xx.x.x",
        "port": 443,
        "uuid": "xxxxxxxxx"
      },
      "tag": "vmess"
    },
    {
      "protocol": "direct",
      "tag": "direct_out"
    }
  ],
  "rules": [
    {
      "external": [
        "site:cn"
      ],
      "target": "direct_out"
    }
  ]
}

direct_out is configured But it didn't work
All traffic still go to failover
what is the problem？

**Reply by mokhtarabadi** [reply]
did you have site.dat file?
if have, please send leaf logs when you run leaf

**Reply by nahddw** [reply]
Site.dat file exists
[2021-06-23 14:47:04][INFO] inbound listening tcp 127.0.0.1:1088
[2021-06-23 14:47:04][INFO] inbound listening udp 127.0.0.1:1088
[2021-06-23 14:47:11][INFO] [] [tcp] [failover_out] [1300ms] ip.cn:443
[2021-06-23 14:47:11][INFO] [] [tcp] [failover_out] [1359ms] www.baidu.com:443
[2021-06-23 14:47:11][INFO] [] [tcp] [failover_out] [1306ms] www.google.com:443
[2021-06-23 14:47:11][INFO] [] [tcp] [failover_out] [1350ms] www.google.com:443
[2021-06-23 14:51:19][INFO] [] [tcp] [failover_out] [1309ms] qq.com:80
[2021-06-23 14:51:19][INFO] [] [tcp] [failover_out] [1309ms] qq.com:80
[2021-06-23 14:51:22][INFO] [] [tcp] [failover_out] [1487ms] www.qq.com:443
[2021-06-23 14:51:25][INFO] [] [tcp] [failover_out] [1287ms] mat1.gtimg.com:443
[2021-06-23 14:51:26][INFO] [] [tcp] [failover_out] [1400ms] mat1.gtimg.com:443
[2021-06-23 14:51:26][INFO] [] [tcp] [failover_out] [1401ms] vm.gtimg.cn:443
[2021-06-23 14:51:26][INFO] [] [tcp] [failover_out] [1400ms] mat1.gtimg.com:443
[2021-06-23 14:51:26][INFO] [] [tcp] [failover_out] [1401ms] mat1.gtimg.com:443
[2021-06-23 14:51:26][INFO] [] [tcp] [failover_out] [1400ms] mat1.gtimg.com:443
[2021-06-23 14:51:29][INFO] [] [tcp] [failover_out] [1285ms] mat1.gtimg.com:443
[2021-06-23 14:51:31][INFO] [] [tcp] [failover_out] [1297ms] img1.gtimg.com:443
[2021-06-23 14:51:31][INFO] [] [tcp] [failover_out] [1310ms] ra.gtimg.com:443

**Reply by nahddw** [reply]
If I use the conf format to run the routing everything is normal.

**Reply by mokhtarabadi** [reply]
change your config to this:
```json
{
  "log": {
    "level": "info"
  },
  "dns": {
    "servers": [
      "1.1.1.1",
      "8.8.8.8"
    ]
  },
  "inbounds": [
    {
      "address": "127.0.0.1",
      "port": 1088,
      "protocol": "socks"
    }
  ],
  "outbounds": [
    {
      "protocol": "failover",
      "settings": {
        "actors": [
          "vmess_out"
        ]
      },
      "tag": "failover_out"
    },
    {
      "protocol": "chain",
      "settings": {
        "actors": [
          "vmess_tls",
          "vmess_ws",
          "vmess"
        ]
      },
      "tag": "vmess_out"
    },
    {
      "protocol": "tls",
      "settings": {
        "serverName": "xxxx",
        "alpn": [
          "http/1.1"
        ]
      },
      "tag": "vmess_tls"
    },
    {
      "protocol": "ws",
      "settings": {
        "path": "/path",
        "headers": {
          "Host": "xx.xx.xx.x"
        }
      },
      "tag": "vmess_ws"
    },
    {
      "protocol": "vmess",
      "settings": {
        "address": "xx.xx.x.x",
        "port": 443,
        "uuid": "xxxxxxxxx"
      },
      "tag": "vmess"
    },
    {
      "protocol": "direct",
      "tag": "direct_out"
    }
  ],
  "router": {
    "domainResolve": false,
    "rules": [
      {
        "external": [
          "site:cn"
        ],
        "target": "direct_out"
      }
    ]
  }
}
```

documentation not completed yet!
in json config need add rules as:
```json
"router": {
    "domainResolve": false,
    "rules": []
  }
```

**Reply by nahddw** [reply]
[INFO] [] [tcp] [direct_out] [104ms] ip111.cn:80
[INFO] [] [tcp] [failover_out] [1349ms] www.google.com:443
It's all right. Great. Thank you, bro

**Reply by mokhtarabadi** [reply]
Good luck bro

**Reply by nahddw** [reply]
@mokhtarabadi 
Bro
How do I use JSON to run on aleaf?

**Reply by mokhtarabadi** [reply]
@nahddw , (I think aleaf project just an demonstrating to releized how leaf works in android), but it's simple
change these below items in SimpleVpnService.kt file
`var configFile = File(filesDir, "config.conf")` to `var configFile = File(filesDir, "config.json")`

and change replace configContent variable contents with

```json
{
  "dns": {
    "servers": [
      "223.5.5.5"
    ]
  },
  "inbounds": [
    {
      "protocol": "tun",
      "settings": {
        "fd": REPLACE-ME-WITH-THE-FD
      },
      "tag": "tun_in"
    }
  ],
  "log": {
    "level": "trace"
  },
  "outbounds": [
    {
      "protocol": "direct",
      "tag": "direct_out"
    }
  ]
}
```
